src/plugins/shape-generator/index.ts

```diff
@@ -0,0 +1,407 @@
+        const wHelper = this.calc.HelperExpression({
+          latex: "w_{rectangleGenerator}",
+        });
+        const hHelper = this.calc.HelperExpression({
+          latex: "h_{rectangleGenerator}",
+        });
+        const angleHelper = this.calc.HelperExpression({
+          latex: "A_{rectangleGenerator}",
+        });
+
+        angleHelper.observe("numericValue", () => {
+          const x = xHelper.numericValue;
+          const y = yHelper.numericValue;
+          const w = wHelper.numericValue;
+          const h = hHelper.numericValue;
+          const angle = angleHelper.numericValue;
+
+          if (isNaN(x) || isNaN(y) || isNaN(w) || isNaN(h) || isNaN(angle)) {
+            return;
+          }
+
+          // TODO: HelperExpression is missing unobserve types
+          angleHelper.unobserve("numericValue");
+
+          // Create new empty expression to add the rectangle to
+          this.calc.controller.dispatch({
+            type: "new-expression",
+          });
+
+          // Generate rectangle LaTeX
+          const latex = rectangleLatex(w, h, x, y, angle);
+
+          // Set the expression
+          this.calc.setExpression({
+            ...expr,
+            id: this.calc.selectedExpressionId,
+            type: "expression",
+            latex,
+          });
+
+          this.cleanupExpressions();
+          this.isEditingShape = false;
+        });
+      },
+    },
+  ];
+
+  private expressionsMutationObserver: MutationObserver | null = null;
+
+  afterEnable() {
+    // Construct stylesheet
+    const style = document.createElement("style");
+    style.textContent = `
+      /* Disable the save button while editing a shape */
+      [data-shape-generator-is-editing-shape="true"] .save-btn-container .dcg-action-save {
+        opacity: 0.5;
+        pointer-events: none;
+      }
+
+      /* Disable pointer events on disabled Add expression dropdown items */
+      .dcg-add-expression-dropdown div.dcg-popover-interior .dcg-new-item.dcg-disabled {
+        pointer-events: none;
+      }
+
+      /* Hide shape points (controls) and helper functions expressions */
+      .dcg-expressionitem[expr-id^="shape-generator-"][expr-id$="-point"],
+      .dcg-expressionitem[expr-id^="shape-generator-"][expr-id*="-helper"] {
+        display: none !important;
+      }
+
+      /* Hide sliders if the setting is disabled */
+      [data-shape-generator-show-sliders="false"] .dcg-expressionitem[expr-id^="shape-generator-"]:has(.dcg-slider) {
+        display: none !important;
+      }
+
+      .shape-generator-ok-btn {
+        margin-top: 36px;
+        margin-left: 1px;
+      }
+    `;
+    style.id = "shape-generator-styles";
+    document.head.appendChild(style);
+
+    this._onDispatchHandler = this.calc.controller.dispatcher.register(
+      this.onDispatch.bind(this)
+    );
+
+    // Listen for expression to add the OK button to
+    this.expressionsMutationObserver = new MutationObserver((mutations) => {
```

Comment:
I don't personally like MutationObserver stuff because they make HTML the source of truth, and we can usually hook into the JS. But if it works it works.

Reply:
I'd typically use .replacements for this, but those are hard to learn and perhaps a little more fragile than MutationObserver (since MutationObserver would break if the HTML output changes, but the .replacements would break if either the HTML output changes or if the underlying DCGView implementation changes).

Reply:
So do you think I should use replacements here? MutationObserver seems to work just fine, and I have no idea how these replacements would work, although I could have a look at other plugins if I had to use them.